common/persistence/elasticsearch/esProcessor.go

```diff
@@ -197,42 +196,79 @@ func (p *esProcessorImpl) bulkAfterAction(_ int64, requests []elastic.BulkableRe
 	if err != nil {
 		// This happens after configured retry, which means something bad happens on cluster or index
 		// When cluster back to live, processor will re-commit those failure requests
-		p.logger.Error("Error commit bulk request.", tag.Error(err))
 
+		isRetryable := isRetryableError(err)
+		p.logger.Error("Unable to commit bulk ES request.", tag.Error(err), tag.Bool(isRetryable))
 		for _, request := range requests {
 			p.logger.Error("ES request failed.", tag.ESRequest(request.String()))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorFailures)
+
+			if !isRetryable {
+				visibilityTaskKey := p.extractVisibilityTaskKey(request)
+				if visibilityTaskKey == "" {
+					continue
+				}
+				p.sendToAckChan(visibilityTaskKey, false)
+			}
 		}
 		return
 	}
 
+	responseIndex := p.buildResponseIndex(response)
 	for i, request := range requests {
-		visibilityTaskKey := p.getVisibilityTaskKey(request)
+		visibilityTaskKey := p.extractVisibilityTaskKey(request)
 		if visibilityTaskKey == "" {
 			continue
 		}
-		responseItem := response.Items[i]
-		for _, resp := range responseItem {
+
+		docID := p.extractDocID(request)
+		if docID == "" {
+			p.sendToAckChan(visibilityTaskKey, false)
+			continue
+		}
+
+		responseItems, ok := responseIndex[docID]
+		if !ok {
+			p.logger.Error("ES request failed. Request item doesn't have corresponding response item.",
+				tag.Value(i),
+				tag.Key(visibilityTaskKey),
+				tag.ESDocID(docID),
+				tag.ESRequest(request.String()))
+			p.sendToAckChan(visibilityTaskKey, false)
+			continue
+		}
+
+		for _, responseItem := range responseItems {
```

Comment:
[Nit] the for loop in the old code ranged over responseItem. Now it ranges over responseItems. Was the previous code incorrectly named?

Reply:
also, what are the scenarios where a responseItems entry has multiple items?

Reply:
Fixed this by merging response items for the same document Id. Added comment to describe logic.